src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs

```diff
@@ -88,7 +99,7 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
                 }
 
                 _contentRemaining = length;
-                // If there is no content then state is still ready
+                // If there is no content then state is reset to ready.
                 _state = _contentRemaining > 0 ? ResponseState.Content : ResponseState.Ready;
                 return 5;
```

Comment:
I know this isn't new, but this seems really weird to me. Especially given the above `// Should never get here. Client always passes 5 to read the header` comment. If this is intended to be consumed in a special way and not as an abstract Stream where I can pass any buffer length, `GrpcWebResponseStream` should not be implementing `Stream`.

Instead, I think it would be better to implement `ReadMessageAsync` directly here in a non-Stream type rather than as an extension method on `StreamExtensions`. That way you don't end up having to do things like parse the header twice. The extension method does seem resilient enough to handle the inner stream returning a partial header just like `TryReadDataAsync` in this `GrpcWebResponseStream` does, so there seems to be a lot of redundant logic.

Reply:
The stream is combined with a custom `HttpContent`, which intercepts the incoming HTTP response body and transforms it. Realistically it will only ever be used by Grpc.Net.Client, which always passes a buffer larger than 5 bytes (unless a zero byte read).

But you're right that it doesn't completely conform to the `Stream` spec. I've updated it to accept any size buffer.